pkg/sql/sqlstats/sslocal/sslocal_provider.go

```diff
@@ -0,0 +1,345 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package sslocal
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/settings"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlstats"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/metric"
+	"github.com/cockroachdb/cockroach/pkg/util/mon"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/pkg/errors"
+)
+
+// NewSQLStatsLocalProvider returns an local in-memory implementation of
+// sqlstats.Provider.
+// TODO(azhng): reportedSQLStats is an unfortunate hack to implement telemetry
```

Comment:
This comment goes with the other one on appStats. 

One thing I'd consider here is that you can have interfaces that are confined to this layer too. We need the `reportedProvider` only really to the `appStats` into it for each app when resetting. My inclination is to expose `appStats` as a new concept in its own package. I'd also move a lot of the iteration business logic over to that package. Like all the stuff that happens on a per-appStat basis could be exposed there. 

I'm not convinced the txnStats concept even needs to leak into this layer.